### ISSUE TITLE
How I can get block size of buffer

### ISSUE THREAD

**Issue by okadabasso** [therealmagicmike/PBKDF2.NET#2]
Hello

I'm trying to generate hash by PBKDF2 with configuration.
In the future I will change hashName in configuration from "HMACSHA256" to "HMACSHA512" or more long sized method.
but it seems to no method to get block size, so I have to write buffer size by literal in my code like  this:

```
 PBKDF2 p = new PBKDF2(password);
byte[] buffer = p.GetBytes(32); // or 64 when HMACSHA512, or 128 or more when hash method in the future

```

I want to write like this:

```
 PBKDF2 p = new PBKDF2(password);
byte[] buffer = p.GetBytes(p.BufferSize);
```

I tried this:

```
var config = ConfigurationManager.GetSection("pbkdf2") as PBKDF2Section;
var algorithm = KeyedHashAlgorithm.Create(config.HashName);
var hashSize = algorithm.HashSize / 8;
 PBKDF2 p = new PBKDF2(password);
byte[] buffer = p.GetBytes(hashSize);
```

It's work well, but I think it is not good because the library is doing same "Hmac.Create" and divide size by 8(1 octet).

Is there any way to  get block size of buffer?

**Comment by therealmagicmike** [comment]
Sorry for the delay in response. I've been busy with other projects and haven't revisited this in a while.

The **GetBytes(int cb)** method's **cb** parameter is intended to be an arbitrary size as per PBKDF2 implementation. The standard defines the requirements for the output length to be at most **(2^32 - 1) * hLen** bytes in length where hLen is the underlying hash's output size, in bytes.

I have seen it mentioned by few that the PBKDF2 output should be of the same or lesser size as the output size of the underlying hash function due to an issue with the keystretching algorithm but I felt it best to leave this choice up to the user and keep the output size arbitrary as per the standard.

The reason I chose not to expose the actual HMAC-based object itself as a property is to keep things simplified for users and to prevent users from being able to mess with the state of the HMAC object and inadvertently conflicting with the internal state of the PBKDF2 object.

As I said, it's been a while since I revisited this but I'll consider exposing the underlying hash size as a read-only property perhaps. I'll look into some other options for this as well perhaps.

As you've already found, aside from hard-coding magic numbers, the only way to get the size of an HMAC object is to get an instance of one.
